### PR TITLE
test_setup: add several functions to handle hugepages

### PR DESCRIPTION
**Pull request by hsmj1412**
Add 2 functions to achieve following:
1.set kernel hugepages
2.get the number of certain hugepages totally allocated by kernel
Update 1 function:
1. get the number of free hugepages for given NUMA node

Signed-off-by: Junxiang Li <junli@redhat.com>

### CODE REVIEW

**Comment by hsmj1412** [comment]
Update from #1289 
REF:
1. avocado.utils.memory and virttest.staging.utils_memory only handle default hugepages size.

below are relevant commands used by similar fuctions in avocado.utils.memory or virttest.staging.utils_memory:
grep Hugepagesize /proc/meminfo --> get default hugepage size
/sbin/sysctl vm.nr_hugepages --> get number of default hugepage size

we can get that those fuctions is only to handle default hugepages size, such as 2048K, so it can't be used to set/get for other huge pages, such as 1G.

So provide more common function to handle each hugepages size.

File: virttest/test_setup.py [hunk 1]
```diff
@@ -577,16 +577,58 @@ def get_multi_supported_hugepage_size(self):
             raise ValueError("Root hugepage control sysfs directory %s did not"
                              " exist" % self.pool_path)
 
-    def get_node_num_huge_pages(self, node, pagesize):
+    def get_kernel_hugepages(self, pagesize):
```
Comment (balamuruhans) [inline virttest/test_setup.py]: please avoid redundant code by abstracting methods `get_node_num_huge_pages()` and `set_node_num_huge_pages()` with your requirement of different sysfs path.

Comment (hsmj1412) [reply virttest/test_setup.py]: @balamuruhans you mean merge the get_node_num_huge_pages and set_node_num_huge_pages? Due to the code have been used before, it maybe affect some existed cases in tp-libvirt and tp-qemu

Comment (balamuruhans) [reply virttest/test_setup.py]: no we should always maintain backward compatibility, I meant to reuse `kernel_interface.py` APIs
```python
def func(self, node_path, value=None):
    hp = kernel_interface.SysFS(node_path)
    if value:
        hp.sys_fs_value = value
        return hp.sys_fs_value == str(value)
     return hp.sys_fs_value
def existing_func(self): # to maintain compatibility
     node_path = "/sys/devices/system/node//hugepages/hugepages-%skB/nr_hugepages"
     return func(node_path)
def new_func(self):
     mem_path = "/sys/kernel/mm/hugepages/hugepages-${pagesize}kB/nr_hugepages"
     return func(mem_path)
```

Comment (hsmj1412) [reply virttest/test_setup.py]: @balamuruhans Updated, please help review again

File: virttest/test_setup.py [hunk 2]
```diff
@@ -577,21 +578,49 @@ def get_multi_supported_hugepage_size(self):
...
+        obj = kernel_interface.SysFS(pagefile)
+        return obj.sys_fs_value
+
+    def set_kernel_hugepages(self, pagesize, pagenum):
+        """
+        Let kernel allocate some specific hugepages at runtime
+         write page number to
+        /sys/kernel/mm/hugepages/hugepages-${pagesize}kB/nr_hugepages
+         :param pagesize: hugepages size
+        :param pagenum: page number
```
Comment (balamuruhans) [inline virttest/test_setup.py]: same here

File: virttest/test_setup.py [hunk 3]
```diff
@@ -577,21 +578,49 @@ def get_multi_supported_hugepage_size(self):
...
-        if not os.path.isfile(node_page_path):
-            raise ValueError("%s page size nr_hugepages file of node %s did "
-                             "not exist" % (pagesize, node))
-        out = decode_to_text(process.system_output("cat %s" % node_page_path))
-        return int(out)
+        node_page_path += "/hugepages/hugepages-%skB/%s" % (pagesize, ptype)
+        obj = kernel_interface.SysFS(node_page_path)
+        return obj.sys_fs_value
 
     def set_node_num_huge_pages(self, num, node, pagesize):
```
Comment (balamuruhans) [inline virttest/test_setup.py]: would you mind fixing it for this method ?

**Comment by balamuruhans** [comment]
@huangyum request for your review. Thanks

File: virttest/test_setup.py (hunk 3 above)
Comment (hsmj1412) [reply virttest/test_setup.py]: @balamuruhans Updated

Comment (huangyum) [reply virttest/test_setup.py]: @hsmj1412  Forgot to push?

Comment (hsmj1412) [reply virttest/test_setup.py]: @huangwei sorry, updated

File: virttest/test_setup.py [hunk 4]
```diff
@@ -577,21 +578,49 @@ def get_multi_supported_hugepage_size(self):
             raise ValueError("Root hugepage control sysfs directory %s did not"
                              " exist" % self.pool_path)
 
-    def get_node_num_huge_pages(self, node, pagesize):
+    def get_kernel_hugepages(self, pagesize):
+        """
+        Get specific hugepages number allocated by kernel at runtime
+        read page number from /sys/kernel/mm/hugepages/hugepages-${pagesize}kB/nr_hugepages
+
+        :param pagesize: hugepages size
```
Comment (huangyum) [inline virttest/test_setup.py]: I guess it's better to explicitly tell the unit of pagesize should be KB.

File: virttest/test_setup.py [hunk 5]
```diff
@@ -603,11 +632,9 @@ def set_node_num_huge_pages(self, num, node, pagesize):
...
         node_page_path = "%s/node%s" % (self.sys_node_path, node)
         node_page_path += "/hugepages/hugepages-%skB/nr_hugepages" % pagesize
-        if not os.path.isfile(node_page_path):
-            raise ValueError("%s page size nr_hugepages file of node %s did "
-                             "not exist" % (pagesize, node))
-        process.system("echo %s > %s" % (num, node_page_path), shell=True)
-        if int(num) != self.get_node_num_huge_pages(node, pagesize):
+        obj = kernel_interface.SysFS(node_page_path)
+        obj.sys_fs_value = num
+        if obj.sys_fs_value != int(num):
```
Comment (huangyum) [inline virttest/test_setup.py]: `num` should always be int.  Would you please correct the parameter description as well, for both set_ and get_ function, thanks.

File: virttest/test_setup.py (hunk 4 above)
Comment (hsmj1412) [reply virttest/test_setup.py]: @huangyum Updated

File: virttest/test_setup.py (hunk 5 above)
Comment (hsmj1412) [reply virttest/test_setup.py]: Due to some num maybe get a str from the config, and the coder forget to turn into int.

**Comment by balamuruhans** [comment]
@hsmj1412 also please have commit msg much more clear with the changes that are done. Let us get the patch in after the review comments are addressed.

**Comment by hsmj1412** [comment]
@balamuruhans @huangyum Updated

**Comment by balamuruhans** [comment]
Thanks @huangyum for review and testing, can we merge it ?

**Comment by huangyum** [comment]
@balamuruhans Yes, I think so. Thanks.

**Pull request merged**